lua/nvim-treesitter/query.lua

```diff
@@ -34,10 +31,42 @@ M.query_extensions = {
   tsx = {'javascript.jsx'}
 }
 
+-- Creates a function that checks whether a certain query exists
+-- for a specific language.
+local function get_query_guard(query)
+  return function(lang)
+    return M.get_query(lang, query) ~= nil
+  end
+end
+
 M.has_locals = get_query_guard('locals')
 M.has_textobjects = get_query_guard('textobjects')
 M.has_highlights = get_query_guard('highlights')
 
+local function read_query_files(filenames)
+  local contents = {}
+
+  for _,filename in ipairs(filenames) do
+    vim.list_extend(contents, vim.fn.readfile(filename))
+  end
+
+  return table.concat(contents, '\n')
+end
+
+local function update_cached_matches(bufnr, changed_tick, query_group)
+  query_cache[query_group][bufnr] = {tick=changed_tick, cache=( M.collect_group_results(bufnr, query_group) or {} )}
```

Comment:
My worry is that a given buffer can have multiple queries of the same kind (for injection purposes mainly), and thus this we not be viable in the long term.

Reply:
Then we could just an additional key for the parser range or just do bit more work on re-collecting the query_group for all parsers.